### PR TITLE
Big June 2021 PR 

### PR DESCRIPTION
**Pull request by AndreaCensi**
Mother issue: https://ethidsc.atlassian.net/browse/DTSW-990

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -4,25 +4,34 @@ ARG ARCH=amd64
...
          gcc \
          libc-dev\
          git \
          bzip2 \
          python3-tk \
          python3-wheel \
          python3-pip  \
+         libcairo2-dev \
+         libjpeg-dev\
+          libgif-dev\
```
Comment (afdaniele) [inline Dockerfile:30]: @AndreaCensi format line

File: Dockerfile [hunk 2]
```diff
@@ -34,34 +43,29 @@ RUN mkdir -p /data/config
 # TODO this is just for the default.yamls - these should really be taken from init_sd_card
 RUN git clone https://github.com/duckietown/duckiefleet.git /data/config
 
+ARG PIP_INDEX_URL="https://pypi.org/simple"
 
-ARG PIP_INDEX_URL
-ENV PIP_INDEX_URL=${PIP_INDEX_URL}
-RUN echo PIP_INDEX_URL=${PIP_INDEX_URL}
+RUN echo we have PIP_INDEX_URL=${PIP_INDEX_URL} $PIP_INDEX_URL $DOCKER_REGISTRY
```
Comment (afdaniele) [inline Dockerfile:48]: @AndreaCensi  do we need this?

File: Dockerfile [hunk 3]
```diff
@@ -34,34 +43,29 @@ RUN mkdir -p /data/config
...
-# Before installing
-RUN echo PYTHONPATH=$PYTHONPATH
-RUN pip3 install -U "pip>=20.2" pipdeptree
 
-RUN pipdeptree
-RUN pip list
+ENV PIP_INDEX_URL=${PIP_INDEX_URL}
+
+RUN echo we have PIP_INDEX_URL=${PIP_INDEX_URL} $PIP_INDEX_URL
+RUN env
```
Comment (afdaniele) [inline Dockerfile:54]: @AndreaCensi  do we need these?

File: Dockerfile [hunk 4]
```diff
@@ -34,34 +43,29 @@ RUN mkdir -p /data/config
...
 
-# FIXME ros-commons is broken
-RUN apt-get update && apt-get install -y libcairo2-dev libjpeg-dev libgif-dev
 
-RUN pip3 install pycairo==1.19.1
-RUN pip check
+#RUN python3 -m pip check # XXX: fails
```
Comment (afdaniele) [inline Dockerfile:57]: remove

File: Dockerfile [hunk 5]
```diff
@@ -34,34 +43,29 @@ RUN mkdir -p /data/config
...
+RUN python3 -m pip list
 
 COPY requirements.* ./
 RUN cat requirements.* > .requirements.txt
-RUN  pip3 install --use-feature=2020-resolver -r .requirements.txt
+RUN python3 -m pip install --no-cache-dir -r .requirements.txt
+RUN python3 -m pip check
+RUN python3 -m pip list
```
Comment (afdaniele) [inline Dockerfile:64]: @AndreaCensi merge into a single `RUN` statement, reduces the number of layers

File: Dockerfile [hunk 6]
```diff
@@ -34,34 +43,29 @@ RUN mkdir -p /data/config
...
 
-RUN echo PYTHONPATH=$PYTHONPATH
-RUN pipdeptree
-RUN pip list
 
 # For ROS Agent - Need to upgrade Pillow for Old ROS stack
-#RUN pip3 install pillow --user --upgrade
+#RUN python3 -m pip install pillow --user --upgrade
```
Comment (afdaniele) [inline Dockerfile:68]: remove

File: submission_ws/src/agent/include/rosagent/rosagent.py [hunk 7]
```diff
@@ -64,83 +67,97 @@ def __init__(self):
...
+        fh = logging.FileHandler('/challenges/challenge-solution-output/rosagent-after-init_node.log')
+        fh.setLevel(logging.DEBUG)
+        # create console handler with a higher log level
+        ch = logging.StreamHandler()
+        ch.setLevel(logging.DEBUG)
+        # create formatter and add it to the handlers
+        # formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+        # fh.setFormatter(formatter)
+        # ch.setFormatter(formatter)
+        # add the handlers to the logger
```
Comment (afdaniele) [inline submission_ws/src/agent/include/rosagent/rosagent.py:81]: @AndreaCensi remove?

File: submission_ws/src/agent/include/rosagent/rosagent.py [hunk 8]
```diff
@@ -64,83 +67,97 @@ def __init__(self):
...
+        root = logging.getLogger()
+        root.addHandler(fh)
+        root.addHandler(ch)
+        #
+        rospy.loginfo('Just after init_node.')
+        # logger.info('my message 2')
+        # rospy.loginfo('Just after init_node [end]')
+        # rospy.logerr('Errors still show up')
```
Comment (afdaniele) [inline submission_ws/src/agent/include/rosagent/rosagent.py:89]: remove

File: submission_ws/src/agent/src/solution.py [hunk 9]
```diff
@@ -1,64 +1,90 @@
...
-        self.agent._publish_info()
+    def on_received_observations(self, data: DB20ObservationsWithTimestamp, context: Context):
+        camera = data.camera
         odometry = data.odometry
-        self.agent._publish_odometry(
-            odometry.resolution_rad,
-            odometry.axis_left_rad,
-            odometry.axis_right_rad
-        )
+        # context.info(f'received obs camera {camera.timestamp} odometry {odometry.timestamp}')
```
Comment (afdaniele) [inline submission_ws/src/agent/src/solution.py:47]: remove

File: submission_ws/src/agent/src/solution.py [hunk 10]
```diff
@@ -1,64 +1,90 @@
...
+            self.agent.publish_odometry(
+                odometry.resolution_rad,
+                odometry.axis_left_rad,
+                odometry.axis_right_rad,
+                odometry.timestamp
+            )
+            self.last_odometry_timestamp = odometry.timestamp
 
     def on_received_get_commands(self, context: Context, data: GetCommands):
+        # context.info(f'on_received_get_commands')
```
Comment (afdaniele) [inline submission_ws/src/agent/src/solution.py:64]: remove

**Review by afdaniele — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 2 above)
Comment (AndreaCensi) [reply Dockerfile:48]: debugging

File: Dockerfile (hunk 3 above)
Comment (AndreaCensi) [reply Dockerfile:54]: debugging

**Pull request merged**